File: ts/src/scanner.ts
import { Token } from "./token.js";
import { TokenType } from "./token_type.js";
import { error } from "./main.js";


export class Scanner {
    src: string;
    tokens: Token[];
    start: number;
    current: number;
    line: number;
    keywords: Map<string, TokenType>;

    constructor(src: string) {
        this.src = src;
        this.tokens = [];
        this.start = 0;
        this.current = 0;
        this.line = 0;

        this.keywords = new Map();
        this.keywords.set("and", TokenType.AND);
        this.keywords.set("class", TokenType.CLASS);
        this.keywords.set("else", TokenType.ELSE);
        this.keywords.set("false", TokenType.FALSE);
        this.keywords.set("for", TokenType.FOR);
        this.keywords.set("fun", TokenType.FUN);
        this.keywords.set("if", TokenType.IF);
        this.keywords.set("nil", TokenType.NIL);
        this.keywords.set("or", TokenType.OR);
        this.keywords.set("print", TokenType.PRINT);
        this.keywords.set("return", TokenType.RETURN);
        this.keywords.set("super", TokenType.SUPER);
        this.keywords.set("this", TokenType.THIS);
        this.keywords.set("true", TokenType.TRUE);
        this.keywords.set("var", TokenType.VAR);
        this.keywords.set("while", TokenType.WHILE);
        this.keywords.set("continue", TokenType.CONTINUE);
        this.keywords.set("break", TokenType.BREAK);
        this.keywords.set("div", TokenType.DIV);
    }

    isAtEnd(): boolean {
        return this.src.length <= this.current;
    }

    advance(): string {
        const char = this.src[this.current];
        this.current += 1;
        return char;
    }

    addToken(type: TokenType, literal?: string | number): void {
        const lexeme = this.src.substring(this.start, this.current);
        let token = new Token(type, lexeme, literal ?? null, this.line);
        this.tokens.push(token);
    }

    match(char: string): boolean {
        if (this.isAtEnd()) return false;
        if (this.src[this.current] !== char) return false;
        this.current += 1;
        return true;
    }

    peek(): string {
        if (this.isAtEnd()) return "\0";
        return this.src[this.current];
    }

    peekNext(): string {
        if (this.src.length <= this.current + 1) return "\0";
        return this.src[this.current + 1];
    }

    string() {
        while (this.peek() != '"' && !this.isAtEnd()) {
            if (this.peek() === "\n") this.line += 1;
            this.advance();
        }

        if (this.isAtEnd()) {
            error(this.line, "Unterminated string");
            return;
        }

        this.advance();
        const str = this.src.substring(this.start + 1, this.current - 1);
        this.addToken(TokenType.STRING, str);
    }

    isDigit(c: string): boolean {
        return "0" <= c && c <= "9";
    }

    isAlpha(c: string): boolean {
        return ("a" <= c && c <= "z") ||
            ("A" <= c && c <= "Z") ||
            c === "_";
    }

    isAlphaNumeric(c: string): boolean {
        return this.isDigit(c) || this.isAlpha(c);
    }

    number() {
        while (this.isDigit(this.peek())) this.advance();

        if (this.peek() === "." && this.isDigit(this.peekNext())) {
            this.advance();
            while (this.isDigit(this.peek())) this.advance();
        }

        this.addToken(TokenType.NUMBER, parseFloat(this.src.substring(this.start, this.current)));
    }

    identifier() {
        while (this.isAlphaNumeric(this.peek())) this.advance();
        const text = this.src.substring(this.start, this.current);
        const type = this.keywords.get(text) ?? TokenType.IDENTIFIER;
        this.addToken(type);
    }

    scanToken(): Token | null {
        const c = this.advance();
        switch (c) {
            case "?": this.addToken(TokenType.QUESTION); break;
            case ":": this.addToken(TokenType.COLON); break;
            case "(": this.addToken(TokenType.LEFT_PAREN); break;
            case ")": this.addToken(TokenType.RIGHT_PAREN); break;
            case "{": this.addToken(TokenType.LEFT_BRACE); break;
            case "}": this.addToken(TokenType.RIGHT_BRACE); break;
            case ",": this.addToken(TokenType.COMMA); break;
            case ".": this.addToken(TokenType.DOT); break;
            case "-": this.addToken(TokenType.MINUS); break;
            case "+": this.addToken(TokenType.PLUS); break;
            case ";": this.addToken(TokenType.SEMICOLON); break;
            case "*": this.addToken(TokenType.STAR); break;
            case "!": this.addToken(this.match("=") ? TokenType.BANG_EQUAL : TokenType.BANG); break;
            case "=": this.addToken(this.match("=") ? TokenType.EQUAL_EQUAL : TokenType.EQUAL); break;
            case "<": this.addToken(this.match("=") ? TokenType.LESS_EQUAL : TokenType.LESS); break;
            case ">": this.addToken(this.match("=") ? TokenType.GREATER_EQUAL : TokenType.GREATER); break;
            case "%": this.addToken(TokenType.PERCENT); break;
            case "/":
                if (this.match("/")) {
                    while (this.peek() !== "\n" && !this.isAtEnd()) this.advance();
                } else if (this.match("*")) {
                    let nesting = 1;
                    while (nesting !== 0 && !this.isAtEnd()) {
                        if (this.match("/") && this.match("*")) nesting += 1;
                        if (this.match("*") && this.match("/")) nesting -= 1;
                        this.advance();
                    }
                } else {
                    this.addToken(TokenType.SLASH);
                }
                break;
            case " ":
            case "\r":
            case "\t":
                break;
            case "\n":
                this.line += 1;
                break;
            case '"': this.string(); break;
            default:
                if (this.isDigit(c)) {
                    this.number();
                } else if (this.isAlpha(c)) {
                    this.identifier();
                } else {
                    error(this.line, `unexpected character '${c}'`);
                }
        }

        return null;
    }

    scanTokens(): Token[] {
        while (!this.isAtEnd()) {
            this.start = this.current;
            this.scanToken();
        }

        const token = new Token(TokenType.EOF, "\0", null, this.line);
        this.tokens.push(token);

        return this.tokens;
    }
}

File: zlox/src/scanner.zig
const std = @import("std");

pub const Scanner = struct {
    start: []const u8,
    current: usize,
    line: u32,
};

pub const Token = struct {
    kind: Token_Kind,
    lexeme: []const u8,
    line: u32,
};

const token_kinds_map = std.ComptimeStringMap(Token_Kind, .{
    .{ "and", .And },
    .{ "class", .Class },
    .{ "else", .Else },
    .{ "if", .If },
    .{ "nil", .Nil },
    .{ "or", .Or },
    .{ "print", .Print },
    .{ "return", .Return },
    .{ "super", .Super },
    .{ "var", .Var },
    .{ "whlie", .While },
});

pub fn create(src: []const u8) Scanner {
    return .{
        .start = src,
        .current = 0,
        .line = 1,
    };
}

pub fn scan_token(s: *Scanner) Token {
    skip_whitespace(s);
    s.start = s.start[s.current..];
    s.current = 0;

    if (is_at_end(s.*)) return make_token(s, .Eof);

    const c = advance(s);

    if (std.ascii.isDigit(c)) return read_number(s);
    if (is_alpha(c)) return read_ident(s);

    switch (c) {
        '(' => return make_token(s, .Left_Paren),
        ')' => return make_token(s, .Right_Paren),
        '{' => return make_token(s, .Left_Brace),
        '}' => return make_token(s, .Right_Brace),
        ';' => return make_token(s, .Semicolon),
        ',' => return make_token(s, .Comma),
        '.' => return make_token(s, .Dot),
        '-' => return make_token(s, .Minus),
        '+' => return make_token(s, .Plus),
        '/' => return make_token(s, .Slash),
        '*' => return make_token(s, .Star),
        '!' => return make_token(s, if (match(s, '=')) .Bang_Equal else .Bang),
        '=' => return make_token(s, if (match(s, '=')) .Equal_Equal else .Equal),
        '<' => return make_token(s, if (match(s, '=')) .Less_Equal else .Less),
        '>' => return make_token(s, if (match(s, '=')) .Greater_Equal else .Greater),
        '"' => return read_string(s),
        else => unreachable,
    }

    return error_token(s, "Unexpected character.");
}

fn is_alpha(c: u8) bool {
    return 'a' <= c and c <= 'z' or 'A' <= c and c <= 'Z' or c == '_';
}

fn read_ident(s: *Scanner) Token {
    while (is_alpha(peek(s.*)) or std.ascii.isDigit(peek(s.*))) _ = advance(s);
    return make_token(s, ident_kind(s));
}

fn ident_kind(s: *Scanner) Token_Kind {
    return token_kinds_map.get(s.start[0..s.current]) orelse .Identifier;
}

fn read_number(s: *Scanner) Token {
    while (std.ascii.isDigit(peek(s.*))) _ = advance(s);

    if (peek(s.*) == '.' and std.ascii.isDigit(peek_next(s.*))) {
        _ = advance(s);

        while (std.ascii.isDigit(peek(s.*))) _ = advance(s);
    }

    return make_token(s, .Number);
}

fn read_string(s: *Scanner) Token {
    while (peek(s.*) != '"' and !is_at_end(s.*)) {
        if (peek(s.*) != '\n') s.line += 1;
        _ = advance(s);
    }

    if (is_at_end(s.*)) return error_token(s, "Unterminated string");

    _ = advance(s);
    return make_token(s, .String);
}

fn skip_whitespace(s: *Scanner) void {
    while (true) {
        const c = peek(s.*);
        switch (c) {
            ' ', '\r', '\t' => _ = advance(s),
            '/' => {
                if (peek_next(s.*) == '/') {
                    while (peek(s.*) != '\n' and !is_at_end(s.*)) _ = advance(s);
                } else {
                    return;
                }
            },
            '\n' => {
                s.line += 1;
                _ = advance(s);
            },
            else => return,
        }
    }
}

fn peek_next(s: Scanner) u8 {
    if (is_at_end(s)) return 0;
    return s.start[s.current + 1];
}

fn peek(s: Scanner) u8 {
    if (is_at_end(s)) return 0;
    return s.start[s.current];
}

fn match(s: *Scanner, expected: u8) bool {
    if (is_at_end(s.*)) return false;
    if (s.current != expected) return false;
    s.current += 1;
    return true;
}

pub fn advance(s: *Scanner) u8 {
    s.current += 1;
    return s.start[s.current - 1];
}

fn make_token(s: *Scanner, kind: Token_Kind) Token {
    return .{
        .kind = kind,
        .lexeme = s.start[0..s.current],
        .line = s.line,
    };
}

fn error_token(s: *Scanner, msg: []const u8) Token {
    return .{
        .kind = .Error,
        .lexeme = msg,
        .line = s.line,
    };
}

fn is_at_end(s: Scanner) bool {
    return s.start.len <= s.current;
}

pub const Token_Kind = enum {
    // Single-character tokens.
    Left_Paren,
    Right_Paren,
    Left_Brace,
    Right_Brace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens.
    Bang,
    Bang_Equal,
    Equal,
    Equal_Equal,
    Greater,
    Greater_Equal,
    Less,
    Less_Equal,
    // Literals.
    Identifier,
    String,
    Number,
    // Keywords.
    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    Error,
    Eof,
};
